lib/msf/core/post/windows/services.rb

```diff
@@ -181,32 +183,61 @@ def each_service(&block)
   # @todo Rewrite to allow operating on a remote host
   #
   def service_list
-    if load_extapi
-      return session.extapi.service.enumerate
-    else
-      serviceskey = "HKLM\\SYSTEM\\CurrentControlSet\\Services"
-      a =[]
-      services = []
-      keys = registry_enumkeys(serviceskey)
-      keys.each do |s|
-        if a.length >= 10
-          a.first.join
-          a.delete_if {|x| not x.alive?}
+    SERVICES.clear
+    case session.type
+    when "powershell"
+      pscommand = '$services = Get-WmiObject win32_service | ?{$_} | where {($_.pathname -ne $null)}'
+      session.shell_command(pscommand)
+      pscommand = 'foreach ($service in $services) {$service.Name+" :: "+$service.PathName+" :: "+$service.StartName+" :: "+$service.StartMode}'
+      serviceslist = session.shell_command(pscommand)
+      serviceslist.each_line do |line|
+        linestr = line.chop
+        unless linestr == ''
+          case linestr.split(" :: ")[-1]
```

Comment:
You're doing this split a bunch of times. Would be better to call it once and store the result.